Port to Windows / Visual Studio Toolchain on Master Branch

jmoguillansky-gpsw opened the pull request:
Port node.gl to Windows using Visual Studio Toolchain.  

Status:
GL backend working on windows.

Ok:                 210
Expected Fail:      0
Fail:               1

This PR can be broken down into the following features:
build: add windows / msvc support
nodegl: port code to windows
ngl-tools: port code to windows

cboesch-gpsw commented on external/scripts/sync.sh:
```diff
@@ -0,0 +1,23 @@
+OS=$1
```
- script is missing a shebang
- if posix shell, you want `set -xeu`
- this script is getting changed several times in the history, you need to squash the commits: I had to review that file several times for nothing

jmoguillansky-gpsw replied:
Thanks I will update...
Regarding the git history I kept two versions of the file but once we sort out the strategy I'll squash the commits.

jmoguillansky-gpsw replied:
updated.  Using set -xu

cboesch-gpsw commented on external/scripts/sync.sh:
```diff
@@ -0,0 +1,23 @@
+    mkdir -p win64
+
+    mkdir -p $TMP_DIR/pkg-config && wget -nc https://github.com/jmoguillansky-gpsw/gopro-nodegl-thirdparty-binaries/blob/master/external/win64/pkg-config.tgz?raw=true -O $TMP_DIR/pkg-config/pkg-config.tgz -P $TMP_DIR/pkg-config
+    cd win64 && tar xzf $TMP_DIR/pkg-config/pkg-config.tgz; cd -
```
please use a sub-shell instead of `cd ; cd -`

jmoguillansky-gpsw replied:
Thanks, good tip.  I'll update...

jmoguillansky-gpsw replied:
updated

cboesch-gpsw commented on external/scripts/sync.sh:
```diff
@@ -1,18 +1,19 @@
 
-    mkdir -p $TMP_DIR/ffmpeg && wget -nc https://github.com/jmoguillansky-gpsw/gopro-nodegl-thirdparty-binaries/blob/master/external/win64/ffmpeg_x64-windows.tgz?raw=true -O $TMP_DIR/ffmpeg/ffmpeg_x64-windows.tgz -P $TMP_DIR/ffmpeg
-    cd win64 && tar xzf $TMP_DIR/ffmpeg/ffmpeg_x64-windows.tgz; cd -
+    mkdir -p $TMP_DIR/ffmpeg && wget --header "Authorization: token $TOKEN" -nc https://api.github.com/repos/jmoguillansky-gpsw/gopro-pkg/actions/artifacts/28753508/zip -O $TMP_DIR/ffmpeg/ffmpeg_x64-windows.zip -P $TMP_DIR/ffmpeg
```
how long will these artifacts survive?

jmoguillansky-gpsw replied:
3 months, but there's several issues with github action artifacts: 1) require security token, 2) different url for each artifact for each run, thst we don't control.  I think if we decide to host prebuilt binaries we should upload them to an ftp server (via github action)

jmoguillansky-gpsw replied:
updated: using binaries from git repo.  They should survive forever

cboesch-gpsw commented on Makefile:
```diff
@@ -154,6 +154,8 @@ ifneq ($(TARGET_OS),MinGW-w64)
 else
 	cp -r $< $@
 endif
+#TODO: submit PR for sxplayer changes
```
can you do that instead? the current branch can't be reviewed until the associated PR is merged in sxplayer

jmoguillansky-gpsw replied:
Lets discuss the strategy first regarding 3rdparty libs.  If everyone agrees, I'll submit PR

jmoguillansky-gpsw replied:
PR submitted

cboesch-gpsw commented on Makefile:
```diff
@@ -29,10 +27,22 @@ PYTHON_MAJOR = 3
 TAR        ?= tar
 TARGET_OS  ?= $(shell uname -s)
 
+ifeq ($(TARGET_OS),Windows)
+PYTHON     ?= python.exe
+PREFIX     ?= nodegl-env
+W_PWD = $(shell wslpath -w .)
+W_PREFIX ?= $(W_PWD)\$(PREFIX)
+$(info PYTHON: $(PYTHON))
+$(info PREFIX: $(PREFIX))
+$(info W_PREFIX: $(W_PREFIX))
+else
```
So you are able to run this Makefile? why the external sync script then? you can download from here

jmoguillansky-gpsw replied:
So we can ignore the details / complexity of the main makefile and develop this code in isolation.  We can put this in a gnu makefile format but I think a bash script is simpler?

cboesch-gpsw commented on Makefile:
```diff
@@ -45,14 +55,28 @@ endif
 SXPLAYER_VERSION ?= 9.6.0
 
+ifeq ($(TARGET_OS), Windows)
+#TODO: identify correct path
```
what isn't correct?

jmoguillansky-gpsw replied:
The sdk path was hardcoded (C:/...)

cboesch-gpsw commented on libnodegl/meson.build:
```diff
@@ -314,7 +330,7 @@ foreach gbackend_name, gbackend_cfg : gbackends_cfg
     gbackend_deps += dependency(dep, required: opt_gbackend)
   endforeach
   foreach dep : host_backend_cfg.get('libs', [])
-    gbackend_deps += cc.find_library(dep, required: opt_gbackend)
+    gbackend_deps += cc.find_library(dep, required: opt_gbackend, dirs: cc_lib_dirs)
```
I know this is a hack we have in the vulkan branch, but it is incorrect: if we do that, it replaces the search dir, which means the libraries won't be searched on the default paths anymore, so it will likely fail typically in an independent build system.

jmoguillansky-gpsw replied:
For some reason msvc finds the library but meson doesn't.  I had to add the search paths explicitly.  Alternatively we could just pass the lib names directly as params

cboesch-gpsw replied:
I think that workaround is OK, thanks

cboesch-gpsw commented on libnodegl/meson.build:
```diff
@@ -305,6 +308,19 @@ if host_system in ['darwin', 'iphone']
+  windows_sdk_library_directories = [ windows_sdk_dir + '\\lib\\10.0.18362.0\\um\\x64' ]
+
+  cc_lib_dirs += windows_sdk_library_directories
+  lib_include_dirs += include_directories('..\\external\\win64\\gl\\include')
```
`..` is incorrect, the builddir could be anywhere: the root makefile is a user of an independently working build system, the dependency doesn't go the other way around.

jmoguillansky-gpsw replied:
Meson converts to absolute path.  It actually complained when I tried to pass it an absolute path if I remember correctly

cboesch-gpsw commented on ngl-tools/meson.build:
```diff
@@ -42,9 +42,13 @@ add_project_arguments(
   dependency('libnodegl'),
 ]
+
+if host_system != 'windows'
+  tool_deps += cc.find_library('m', required: false)
+endif
```
why do you need this?

jmoguillansky-gpsw replied:
I can probably remove it...

cboesch-gpsw replied:
yes but `required` is set to `false` so why is it an issue?

jmoguillansky-gpsw replied:
it also makes it more clear to the developer which libraries are required for different platforms.  
I can remove this change if you prefer.

cboesch-gpsw replied:
We're just following the recommended practice from here: https://mesonbuild.com/howtox.html#add-math-library-lm-portably

The purpose of `required: false` really is all about this: we want as little platform specific logic and be feature/availability oriented. Please don't change something that works for something more complex.

cboesch-gpsw commented on ngl-tools/meson.build:
```diff
@@ -115,6 +119,12 @@ tools_specs = {
 }
 
+if host_system == 'windows'
+  tool_install_dir = 'Scripts'
```
"Scripts"? Those are binary programs

jmoguillansky-gpsw replied:
Yes its not the best naming.  When we run python -m venv on windows it sets up a scripts folder but no bin folder.  We can add a folder "bin" but we need to add it to path env variable in the activate script...

cboesch-gpsw commented on libnodegl/node_buffer.c:
```diff
@@ -127,15 +126,16 @@ static int buffer_init_from_filename(struct ngl_node *node)
-    off_t ret      = lseek(s->fd, 0, SEEK_SET);
-    if (filesize < 0 || ret < 0) {
+    fseek(s->fp, 0, SEEK_END);
+    int filesize = ftell(s->fp);
```
this breaks with large files

jmoguillansky-gpsw replied:
Will replace with long int

jmoguillansky-gpsw replied:
already resolved

cboesch-gpsw commented on libnodegl/utils.c:
```diff
@@ -47,20 +51,21 @@ char *ngli_strdup(const char *s)
-{
-    struct timeval tv;
-
-    gettimeofday(&tv, NULL);
-    return 1000000 * (int64_t)tv.tv_sec + tv.tv_usec;
-}
-
 int64_t ngli_gettime_relative(void)
```
This belongs in a dedicated commit, prototype removal is forgotten btw.

jmoguillansky-gpsw replied:
Thanks I'll update...

jmoguillansky-gpsw replied:
updated

cboesch-gpsw commented on libnodegl/backends/gl/glincludes.h:
```diff
@@ -57,17 +57,14 @@
 
-#ifdef _WIN32
-#define NGLI_GL_APIENTRY WINAPI
-#else
 #define NGLI_GL_APIENTRY
```
mmh what's going on here? the commit description needs to explain that

jmoguillansky-gpsw replied:
Lots of compile errors if I keep "winapi"... with this change the gl backend works.  Is "winapi" related to symbol visibility and dlls?

jmoguillansky-gpsw replied:
This introduced compile errors.  No need to do this

cboesch-gpsw replied:
Mmh well then, `NGLI_GL_APIENTRY` needs to be drop entirely: it's sole purpose was to work with windows

jmoguillansky-gpsw replied:
Can you please help with that change?  
If I recall correctly, glfunctions.h is autogenerated, I don't have to modify it by hand?

mbouron-gpsw replied:
NGLI_GL_APIENTRY is still needed. The build issue with MSVC is related to the invalid syntax we currently use (which is  accepted by gcc) for declaring the OpenGL function pointers. https://github.com/gopro/gopro-lib-node.gl/pull/166 should addresses the issue.

jmoguillansky-gpsw replied:
I tried #166 but still get compile errors (see jmoguillansky/meson-windows-master-tmp-gl-apientry)...

cboesch-gpsw commented on libnodegl/api.c:
```diff
@@ -70,7 +70,7 @@ static int get_default_platform(void)
 #elif defined(TARGET_ANDROID)
     return NGL_PLATFORM_ANDROID;
-#elif defined(TARGET_MINGW_W64)
+#elif defined(TARGET_WINDOWS)
```
you can't do that in an isolated commit: it needs to be associated with changes in the `meson.build`, in the CI and in the documentation

jmoguillansky-gpsw replied:
We should define "NGL_PLATFORM_WINDOWS" if TARGET_WINDOWS.  Not sure if we need two versions: "NGL_PLATFORM_WINDOWS and "NGL_PLATFORM_MINGW"?

jmoguillansky-gpsw replied:
We have 3 macros:
TARGET_WINDOWS, TARGET_MINGW_W64, and TARGET_MSVC.
For this particular logic, it can be applied to both msvc and mingw (hence TARGET_WINDOWS)
We have the corresponding changes to meson.build and CI.

cboesch-gpsw commented on ngl-tools/ipc.c:
```diff
@@ -22,10 +22,11 @@
-#include <unistd.h>
 #ifdef _WIN32
 #include <winsock2.h>
+#define ssize_t int
```
why do you define `ssize_t` if you remove it from the code? also, this is defined in mingw, but not msvc apparently

jmoguillansky-gpsw replied:
It was either define it here or put ifdefs everywhere that recv / send is called and use int on windows and ssize _t on unix.  This was basically a trick to reduce the code changes.  Alternatively we could use macros or abstraction APIs

jmoguillansky-gpsw replied:
already resolved

cboesch-gpsw commented on pynodegl/setup.py:
```diff
@@ -36,13 +36,18 @@ def __init__(self, pkg_config_bin='pkg-config'):
-        self.library_dirs = [f[2:] for f in flags if f.startswith('-L')]
-        self.libraries    = [f[2:] for f in flags if f.startswith('-l')]
-
+        pkgcfg_cflags      = subprocess.check_output([pkg_config_bin, '--cflags', self.PKG_LIB_NAME]).decode()
+        pkgcfg_libs        = subprocess.check_output([pkg_config_bin, '--libs', self.PKG_LIB_NAME]).decode()
+        cflags = pkgcfg_cflags.split()
+        ldflags = pkgcfg_libs.split()
+        self.include_dirs = [f[2:] for f in cflags if f.startswith('-I')]
+        self.library_dirs = [f[2:] for f in ldflags if f.startswith('-L')]
+        self.libraries = []
+        for f in ldflags:
+            if f.startswith('-l'):
+                self.libraries.append(f[2:])
+            elif f.endswith('.lib'):
+                self.libraries.append(f[:-4])
```
This change doesn't belong in `.gitignore` commit

jmoguillansky-gpsw replied:
I'll clean up the history...

jmoguillansky-gpsw replied:
updated

jmoguillansky-gpsw commented on .gitattributes:
```diff
@@ -0,0 +1 @@
+* -text
```
The CI server is converting the line endings (e.g. \n to \r\n) automatically for various files.  I had to disable this

cboesch-gpsw replied:
please write that in the commit description (and add a "git" prefix to the commit message)

cboesch-gpsw replied:
Also, doesn't this affect all binaries? Like, is git going to consider our images in the doc as text now?

jmoguillansky-gpsw replied:
gitattributes(5) (kernel.org)<https://mirrors.edge.kernel.org/pub/software/scm/git/docs/gitattributes.html>

text
This attribute enables and controls end-of-line normalization. When a text file is normalized, its line endings are converted to LF in the repository. To control what line ending style is used in the working directory, use the eol attribute for a single file and the core.eol configuration variable for all text files. Note that setting core.autocrlf to true or input overrides core.eol (see the definitions of those options in git-config(1)<https://mirrors.edge.kernel.org/pub/software/scm/git/docs/git-config.html>).

We’re disabling this feature (the minus in front) so we’re disabling end-of-line normalization. It won’t affect any binaries, etc.  All this does is turn off end-of-line normalization of text files.

ubitux replied:
So it means someone working on windows may start submitting files with CLRF line endings?

Unless I'm mistaken, this is only needed for writing file, so it's not about the clone itself. Why do you need to write files in the first place? Is that related to the OpenGL wrapper thing?

jmoguillansky-gpsw commented on .github/workflows/ci_win.yml:
```diff
@@ -30,3 +30,69 @@ jobs:
+
+      - name: Build
+        run: |
+          make -j8 TARGET_OS=Windows PKG_CONFIG="D:\\\\a\\\\gopro-lib-node.gl\\\\gopro-lib-node.gl\\\\pkg-config.exe" VCPKG_DIR="D:\\\\a\\\\gopro-lib-node.gl\\\\gopro-lib-node.gl\\\\vcpkg"
```
can probably use single quotes here instead of lots of escape characters, still learning bash scripting...
https://wiki.bash-hackers.org/

cboesch-gpsw commented on .gitignore:
```diff
@@ -1,2 +1,5 @@
-builddir
+/builddir
```
this looks unrelated

cboesch-gpsw commented on .gitignore:
```diff
@@ -1,2 +1,5 @@
+/builddir
 nodegl-env/
+/sxplayer*
+/external/win64
+external/token.cfg
```
this is inconsistent and doesn't look like it doesn't belong in this commit but in the one introducing the external w64 stuff

cboesch-gpsw commented on .gitignore:
```diff
@@ -1,2 +1,5 @@
-builddir
+/builddir
 nodegl-env/
+/sxplayer*
```
you can drop that

cboesch-gpsw commented on ngl-tools/meson.build:
```diff
@@ -120,6 +120,8 @@ tools_specs = {
 }
 
 if host_system == 'windows'
+  #Python -m venv on windows creates a Scripts folder instead of a bin folder
+  #TODO: add a bin folder and add it to path in activate script
```
1. this diff doesn't belong in "nodegl: use TARGET_WINDOWS instead of TARGET_MINGW"
2. cosmetic: please add a space after `#`
3. what is the plan wrt the TODO?

jmoguillansky-gpsw replied:
3: We can post a issue to python developers.  This is a python issue.  
For example, meson.exe is also in Scripts folder in python system path:

ls /mnt/c/Users/jmoguillansky/AppData/Local/Programs/Python/Python38/Scripts/

cygdb.exe   cythonize.exe         easy_install.exe  ngl-test-script.py  ngl-viewer-script.py  pip.exe     pip3.exe              pyside2-lupdate.exe  pyside2-uic.exe
cython.exe  easy_install-3.8.exe  meson.exe         ngl-test.exe        ngl-viewer.exe        pip3.8.exe  pyside2-designer.exe  pyside2-rcc.exe      wheel.exe

cboesch-gpsw commented on Makefile:
```diff
@@ -181,6 +181,8 @@ endif
 nodegl-install: nodegl-setup
 ifeq ($(TARGET_OS),Windows)
 	(cmd.exe /C $(ACTIVATE) \&\& $(MESON_COMPILE) -C builddir\\libnodegl \&\& $(MESON_INSTALL) -C builddir\\libnodegl)
+	# patch libnodegl.pc TODO: remove
```
what's your plan for this?

jmoguillansky-gpsw replied:
It seems that currently meson pkg-config export doesn't support lib paths with spaces.  
I will post an issue on meson github...

cboesch-gpsw commented on Makefile:
```diff
@@ -225,13 +225,25 @@ else
 
 tests: nodegl-tests tests-setup
+ifeq ($(TARGET_OS),Windows)
+	(cmd.exe /C $(ACTIVATE) \&\& meson test $(MESON_TESTS_SUITE_OPTS) -C builddir\\tests)
```
if you include `cmd.exe /C` in `ACTIVATE`, does it help getting rid of the condition?

jmoguillansky-gpsw replied:
There's two differences between windows and others:
1) we use cmd.exe
2) we use different slashes (e.g. forward slash instead of backslashes).
It would be great if we can conslidate the code and avoid platform-specific code, but this can be a future improvement.  
It's not so easy to do this in a makefile, easier in a high level programming language like Python, etc.

ubitux replied:
> we use different slashes (e.g. forward slash instead of backslashes).

I though windows was supporting `/` nowadays?

> It's not so easy to do this in a makefile, easier in a high level programming language like Python, etc.

I'm not opposed to having Windows handled outside the Makefile (we moved away from make for the build for a reason). I don't know if Python would be a great language for bootstrapping packages etc on Windows (because it requires an initial installation for the user and we will need to create a virtual env from another Python script, it might get clumsy), but I wouldn't be particularly opposed to that. Isn't Powershell moer appropriate? (I really don't know).

If you were under the impression that I was asking for Windows integration within the Makefile, I'm sorry. The goal of that Makefile is to bootstrap the environment: we kept make because on Unix it's extremely convenient. The keyword here is "convenient", so the question to answer on windows is: "what would be the most convenient way for a user to bootstrap the environment reliably?".

cboesch-gpsw commented on libnodegl/meson.build:
```diff
@@ -344,7 +344,7 @@ foreach gbackend_name, gbackend_cfg : gbackends_cfg
     all_dep_found = all_dep_found and dep.found()
   endforeach
-
+  message('backend: ' + gbackend_name + ' enabled: ' + all_dep_found.to_string())
```
`message('backend: @0@ enabled: @1@'.format(gbackend_name, all_dep_found.to_string())` maybe, for consistency with the other messages we have in other `meson.build` files

jmoguillansky-gpsw replied:
ok I'll update...

cboesch-gpsw commented on Makefile:
```diff
@@ -63,6 +63,7 @@ RPATH_LDFLAGS ?= -Wl,-rpath,$(PREFIX)/lib
 
 ifeq ($(TARGET_OS),Windows)
 MESON_SETUP   = meson setup --backend vs --prefix="$(W_PREFIX)" --pkg-config-path=$(PREFIX)\\Lib\\pkgconfig -Drpath=true
+MESON_SETUP_NINJA   = meson setup --backend ninja --prefix="$(W_PREFIX)" --pkg-config-path=$(PREFIX)\\Lib\\pkgconfig -Drpath=true
```
why is this commit separated from the rest? isn't it needed at the moment windows support is introduced?

jmoguillansky-gpsw replied:
the unit tests target regarding meson setup using ninja.  All the other targets can be built with visual studio backend.

cboesch-gpsw commented on Makefile:
```diff
@@ -200,9 +202,9 @@ endif
 sxplayer-install: external-download $(PREFIX)
 ifeq ($(TARGET_OS),Windows)
-	(cmd.exe /C $(ACTIVATE) \&\& $(MESON_SETUP) --default-library static external\\sxplayer builddir\\sxplayer \&\& $(MESON_COMPILE) -C builddir\\sxplayer \&\& $(MESON_INSTALL) -C builddir\\sxplayer)
+	(cmd.exe /C $(ACTIVATE) \&\& $(MESON_SETUP) --default-library shared external\\sxplayer builddir\\sxplayer \&\& $(MESON_COMPILE) -C builddir\\sxplayer \&\& $(MESON_INSTALL) -C builddir\\sxplayer)
```
same here, it looks like you want this squashed into the windows build commit

cboesch-gpsw commented on libnodegl/meson.build:
```diff
@@ -373,14 +373,6 @@ lib_link_args += cc.get_supported_link_arguments([
-  lib_name_prefix = ''
-  lib_name_suffix = 'lib'
-endif
-
```
same here: i think this belongs in the commit adding windows build support

Note: "update X" is not a good commit message, a commit pretty much always updates something

cboesch-gpsw commented on libnodegl/meson.build:
```diff
@@ -62,8 +62,12 @@ conf_data.set10('DEBUG_GL', 'gl' in debug_opts)
 
-if host_system == 'windows' and cc.get_id() != 'msvc'
-  conf_data.set10('TARGET_MINGW_W64', true)
+if host_system == 'windows'
+  if cc.get_id() == 'msvc'
+    conf_data.set10('TARGET_MSVC', true)
+  else
+    conf_data.set10('TARGET_MINGW_W64', true)
+  endif
```
please rebase, this should be available upstream

cboesch-gpsw commented on Makefile:
```diff
@@ -219,12 +225,12 @@ external-download:
-	(cmd.exe /C pushd external\\win64\\sdl2_x64-windows \&\& python.exe scripts/install.py "$(W_PREFIX)" \& popd)
+	(cmd.exe /C pushd external\\win64\\ffmpeg_x64-windows \&\& $(PYTHON) scripts/install.py "$(W_PREFIX)" \& popd)
+	(cmd.exe /C pushd external\\win64\\pthreads_x64-windows \&\& $(PYTHON) scripts/install.py "$(W_PREFIX)" \& popd)
+	(cmd.exe /C pushd external\\win64\\sdl2_x64-windows \&\& $(PYTHON) scripts/install.py "$(W_PREFIX)" \& popd)
```
again i think this commit belongs in another

cboesch-gpsw commented on libnodegl/utils.c:
```diff
@@ -23,11 +23,14 @@
 #include <pthread.h>
 
 #ifdef _WIN32
+#define POW10_9 1000000000
+#include <Windows.h>
```
> Resolves compile error "No target architecture"

What's going on here? What following include requires `Windows.h` to be included first? Can you add the explanation in the commit description?

Also, we don't mention "bugfix" like this. "utils: fix Windows.h include order" is more consistent with other commits.

jmoguillansky-gpsw replied:
I get the compile error:
C:\Program Files (x86)\Windows Kits\10\Include\10.0.18362.0\um\winnt.h(173,1): fatal error C1189: #error:  "No Target Architecture" [C:\Users\jmoguillansky\Documents\2020\Git\gopro-lib-node.gl-meson-windows-master\builddir\libnodegl\tes
t_utils@exe.vcxproj]

I put in the commit message it fixes compile error "No Target Architecture".  
Is this sufficient detail?

cboesch-gpsw replied:
Yes but why re-ordering fix it? What includes set the architecture, and what is complaining about this missing architecture?

cboesch-gpsw commented on libnodegl/log.c:
```diff
@@ -22,11 +22,11 @@
 #include <stdio.h>
 #include <stdlib.h>
+#include "config.h"
 #if !defined(TARGET_IPHONE) && !defined(TARGET_ANDROID) && !defined(TARGET_WINDOWS)
 #include <unistd.h>
 #endif
 
-#include "config.h"
```
Aw. We need to merge that one quickly. Please mention in the commit description that it completes the initial attempt in 1494a18c11b2466e3c2c89fa133205bc3ac9be90

cboesch-gpsw commented on Makefile:
```diff
@@ -48,7 +48,18 @@ TESTS_SUITE ?=
+endif
+
+ifeq ($(TARGET_OS), Windows)
+  ifneq ($(shell $(CMD) $(PYTHON) -c "import sys;print(sys.version_info.major)"),$(PYTHON_MAJOR))
```
again, please squash all of this in the same commit; we don't need to see all your iteration attempts.

cboesch-gpsw commented on .github/workflows/ci_win.yml:
```diff
@@ -30,3 +30,69 @@ jobs:
+        run: |
+          make -j8 TARGET_OS=Windows PKG_CONFIG="D:\\\\a\\\\gopro-lib-node.gl\\\\gopro-lib-node.gl\\\\pkg-config.exe" VCPKG_DIR="D:\\\\a\\\\gopro-lib-node.gl\\\\gopro-lib-node.gl\\\\vcpkg"
+
+#      - name: Setup debug session using tmate
+#        if: always()
+#        run: |
+#          sudo apt -y update
+#          #Install tmate
+#          sudo apt-get install -y tmate openssh-client
+#          # Generate ssh key if needed
+#          [ -e ~/.ssh/id_rsa ] || ssh-keygen -t rsa -f ~/.ssh/id_rsa -q -N ""
+#          # Run deamonized tmate
+#          echo Running tmate...
+#          tmate -S /tmp/tmate.sock new-session -d
+#          tmate -S /tmp/tmate.sock wait tmate-ready
+#          # Print connection info
+#          tmate -S /tmp/tmate.sock display -p '#{tmate_ssh}'
+#
+#      - name: Wait for connection
+#        if: always()
+#        run: |
+#          # Wait for connection
+#          while [ -S /tmp/tmate.sock ]; do
+#            sleep 1
+#          done
+
+        
+#      - name: Run tests
+#        run: |
+#          make -j8 tests PYTHON="C:/hostedtoolcache/windows/Python/3.9.0/x64/python.exe" TARGET_OS=Windows
```
No dead code in the repository please: put that in your personal notes or write a documentation somewhere

cboesch-gpsw commented on .github/workflows/ci_win.yml:
```diff
@@ -30,3 +30,69 @@ jobs:
+          sudo apt -y update
+          sudo apt -y install build-essential zip
+          pip.exe install meson
+          wget https://github.com/jmoguillansky-gpsw/gopro-pkg/blob/master/external/win64/pkg-config.exe?raw=true -O pkg-config.exe
```
can we have something more official that this?

jmoguillansky-gpsw replied:
Windows doesn't provide a pkg-config windows executable.  
I extracted it from msys2, and put it for now in a personal repo. 
There doesn't really seem to be a standalone windows pre-built binary available anywhere.  
We can either build this from source as part of node.gl build, or we can provide this in a public ftp server (or something equivalent).

jbnoel-gpsw replied:
I agree with @cboesch-gpsw, we need to limit the hacks.
Ideas:
- prebuilt with github action like other dependencies ?
- https://stackoverflow.com/questions/1710922/how-to-install-pkg-config-in-windows ?

jmoguillansky-gpsw replied:
Regarding pkg config: yes I'm currently working on a proper solution but its non trivial.  I was able to build "pkgconf" from source: this is supposed to be a drop in replacememt for pkg config but there's compatibility issues on Windows.  It seems I need to patch the source code too... This was meant as a temporary workaround.

FYI:
when I build with "pkg-config" from msys2:

Called `C:\Users\jmoguillansky\Documents\2020\Git\gopro-lib-node.gl-meson-windows-master\nodegl-env\Scripts\pkg-config.EXE sdl2 --libs` -> 0
-L -L/manual-link -lSDL2 -lSDL2main -ldinput8 -lshell32 -lsetupapi -ladvapi32 -luuid -lversion -loleaut32 -lole32 -limm32 -lwinmm -lgdi32 -luser32 -lSDL2

When I build with "pkgconf" (built from source):
Called `C:\Users\jmoguillansky\Documents\2020\Git\gopro-lib-node.gl-meson-windows-master\nodegl-env\Scripts\pkg-config.EXE sdl2 --libs` -> 0
-L -lSDL2 -L/manual-link -lSDL2main

I need to add param "--static".  There's various other incompatibilities like backslashes instead of forward slashes, etc

jmoguillansky-gpsw commented:
Updated

Get Outlook for Android<https://aka.ms/ghei36>

________________________________
From: cboesch-gpsw <notifications@github.com>
Sent: Wednesday, January 6, 2021 1:15:17 AM
To: gopro/gopro-lib-node.gl <gopro-lib-node.gl@noreply.github.com>
Cc: Jeffrey Moguillansky <jmoguillansky@gopro.com>; Author <author@noreply.github.com>
Subject: Re: [gopro/gopro-lib-node.gl] Port to Windows / Visual Studio Toolchain on Master Branch (#152)


@cboesch-gpsw commented on this pull request.

________________________________

In ngl-tools/meson.build<https://github.com/gopro/gopro-lib-node.gl/pull/152#discussion_r552454256>:

> +if host_system != 'windows'
+  tool_deps += cc.find_library('m', required: false)
+endif


We're just following the recommended practice from here: https://mesonbuild.com/howtox.html#add-math-library-lm-portably

The purpose of required: false really is all about this: we want as little platform specific logic and be feature/availability oriented. Please don't change something that works for something more complex.

—
You are receiving this because you authored the thread.
Reply to this email directly, view it on GitHub<https://github.com/gopro/gopro-lib-node.gl/pull/152#discussion_r552454256>, or unsubscribe<https://github.com/notifications/unsubscribe-auth/AO3BUEYYHAAX7MVZQJW4EJLSYQS2LANCNFSM4UDAE7OA>.

jmoguillansky-gpsw reviewed (commented):
Note that these files were autogenerated on windows platform.  
This is why the line endings are different.

cboesch-gpsw commented on libnodegl/meson.build:
```diff
@@ -491,38 +491,24 @@ endif
-  output: 'gl.xml',
+message('** generate gl wrappers **')
+message('gl.xml path: ' + meson.current_build_dir() / 'gl.xml')
+run_command('curl',
+  'https://raw.githubusercontent.com/KhronosGroup/OpenGL-Registry/master/xml/gl.xml',
+  '-o', meson.current_build_dir() / 'gl.xml'
```
We don't want to do that at every run; please keep the custom target. It doesn't matter if it doesn't work on Windows, just call the script directly if needed.

jmoguillansky-gpsw replied:
how about if we use "run_command" on msvc and "custom_target" for others, and put a comment that this is a temporary solution?

cboesch-gpsw commented on pynodegl/.gitignore:
```diff
@@ -4,3 +4,5 @@ pynodegl.c
 pynodegl.egg-info
 pynodegl.*.so
 .eggs
+*.pyd
+*.dll
```
Again "update X" is not a good commit message; please squash this into the corresponding commit

cboesch-gpsw commented on Makefile:
```diff
@@ -19,36 +19,71 @@
-ifneq ($(shell $(PYTHON) -c "import sys;print(sys.version_info.major)"),$(PYTHON_MAJOR))
-$(error "Python $(PYTHON_MAJOR) not found")
+ifeq ($(TARGET_OS), Windows)
+VCVARS64 ?= "C:\Program Files (x86)\Microsoft Visual Studio\2019\Enterprise\VC\Auxiliary\Build\vcvars64.bat"
+ACTIVATE = $(VCVARS64) \&\& $(PREFIX)\\Scripts\\activate.bat
+VCPKG_DIR ?= C:\\vcpkg
+PKG_CONFIG ?= C:\\msys64\\usr\\bin\\pkg-config.exe
```
Can't we do something about these hardcoded path?

jmoguillansky-gpsw replied:
those are the default params, the user can override

cboesch-gpsw commented on Makefile:
```diff
@@ -119,48 +166,110 @@ pynodegl-utils-install: pynodegl-utils-deps-install
+ifeq ($(TARGET_OS),Windows)
+	($(CMD) $(ACTIVATE) \&\& pip -v install -e .\\pynodegl)
+	# Install DLLs / Libs
+	($(CMD) xcopy /Y builddir\\sxplayer\\*.dll pynodegl\\.)
+	($(CMD) xcopy /Y /C builddir\\libnodegl\\*.dll pynodegl\\.)
+	($(CMD) xcopy /Y builddir\\sxplayer\\*.dll $(PREFIX)\\Scripts\\.)
+	($(CMD) xcopy /Y /C builddir\\libnodegl\\*.dll $(PREFIX)\\Scripts\\.)
+	($(CMD) xcopy /Y /C builddir\\libnodegl\\*.lib $(PREFIX)\\Lib\\.)
```
The `( ... )` form is a subshell and is only needed when you use the "activate", you can get rid of them

jmoguillansky-gpsw commented:
Regarding hardcoded path, thats the default value.  The user can override (and I override it locally because I use community edition)

Get Outlook for Android<https://aka.ms/ghei36>

________________________________
From: cboesch-gpsw <notifications@github.com>
Sent: Thursday, January 7, 2021 8:10:36 AM
To: gopro/gopro-lib-node.gl <gopro-lib-node.gl@noreply.github.com>
Cc: Jeffrey Moguillansky <jmoguillansky@gopro.com>; Author <author@noreply.github.com>
Subject: Re: [gopro/gopro-lib-node.gl] Port to Windows / Visual Studio Toolchain on Master Branch (#152)


@cboesch-gpsw commented on this pull request.

________________________________

In Makefile<https://github.com/gopro/gopro-lib-node.gl/pull/152#discussion_r553426409>:

> +VCVARS64 ?= "C:\Program Files (x86)\Microsoft Visual Studio\2019\Enterprise\VC\Auxiliary\Build\vcvars64.bat"
+ACTIVATE = $(VCVARS64) \&\& $(PREFIX)\\Scripts\\activate.bat
+VCPKG_DIR ?= C:\\vcpkg
+PKG_CONFIG ?= C:\\msys64\\usr\\bin\\pkg-config.exe


Can't we do something about these hardcoded path?

—
You are receiving this because you authored the thread.
Reply to this email directly, view it on GitHub<https://github.com/gopro/gopro-lib-node.gl/pull/152#pullrequestreview-563616092>, or unsubscribe<https://github.com/notifications/unsubscribe-auth/AO3BUE2SFBBMVH4QYYO3SPLSYXMHZANCNFSM4UDAE7OA>.

jmoguillansky-gpsw commented:
Regarding pkg config: yes I'm currently working on a proper solution but its non trivial.  I was able to build "pkgconf" from source: this is supposed to be a drop in replacememt for pkg config but there's compatibility issues on Windows.  It seems I need to patch the source code too... This was meant as a temporary workaround.

Get Outlook for Android<https://aka.ms/ghei36>
________________________________
From: jbnoel-gpsw <notifications@github.com>
Sent: Thursday, January 7, 2021 8:01:59 AM
To: gopro/gopro-lib-node.gl <gopro-lib-node.gl@noreply.github.com>
Cc: Jeffrey Moguillansky <jmoguillansky@gopro.com>; Author <author@noreply.github.com>
Subject: Re: [gopro/gopro-lib-node.gl] Port to Windows / Visual Studio Toolchain on Master Branch (#152)


@jbnoel-gpsw commented on this pull request.

________________________________

In .github/workflows/ci_win.yml<https://github.com/gopro/gopro-lib-node.gl/pull/152#discussion_r553420998>:

> +      run:
+        shell: wsl-bash {0}
+
+    steps:
+      - uses: Vampire/setup-wsl@v1
+        with:
+          distribution: Ubuntu-20.04
+
+      - uses: actions/checkout@v2
+
+      - name: Install dependencies
+        run: |
+          sudo apt -y update
+          sudo apt -y install build-essential zip
+          pip.exe install meson
+          wget https://github.com/jmoguillansky-gpsw/gopro-pkg/blob/master/external/win64/pkg-config.exe?raw=true -O pkg-config.exe


I agree with @cboesch-gpsw<https://github.com/cboesch-gpsw>, we need to limit the hacks.
Ideas:

  *   prebuilt with github action like other dependencies ?
  *   https://stackoverflow.com/questions/1710922/how-to-install-pkg-config-in-windows ?

—
You are receiving this because you authored the thread.
Reply to this email directly, view it on GitHub<https://github.com/gopro/gopro-lib-node.gl/pull/152#discussion_r553420998>, or unsubscribe<https://github.com/notifications/unsubscribe-auth/AO3BUE22PBWZK3ACKERFASLSYXLHPANCNFSM4UDAE7OA>.

jmoguillansky-gpsw commented:
What about using run_command on msvc and custom target on others?

From: cboesch-gpsw <notifications@github.com>
Sent: Thursday, January 7, 2021 8:08 AM
To: gopro/gopro-lib-node.gl <gopro-lib-node.gl@noreply.github.com>
Cc: Jeffrey Moguillansky <jmoguillansky@gopro.com>; Author <author@noreply.github.com>
Subject: Re: [gopro/gopro-lib-node.gl] Port to Windows / Visual Studio Toolchain on Master Branch (#152)


@cboesch-gpsw commented on this pull request.

________________________________

In libnodegl/meson.build<https://github.com/gopro/gopro-lib-node.gl/pull/152#discussion_r553424689>:

> +run_command('curl',

+  'https://raw.githubusercontent.com/KhronosGroup/OpenGL-Registry/master/xml/gl.xml',

+  '-o', meson.current_build_dir() / 'gl.xml'

We don't want to do that at every run; please keep the custom target. It doesn't matter if it doesn't work on Windows, just call the script directly if needed.

—
You are receiving this because you authored the thread.
Reply to this email directly, view it on GitHub<https://github.com/gopro/gopro-lib-node.gl/pull/152#pullrequestreview-563613665>, or unsubscribe<https://github.com/notifications/unsubscribe-auth/AO3BUE6T2XYBTS3OIKQTPDTSYXL5FANCNFSM4UDAE7OA>.

jmoguillansky-gpsw commented:
Winnt.h which indirectly included by some other include…
Not so easy to trace the culprit

From: cboesch-gpsw <notifications@github.com>
Sent: Thursday, January 7, 2021 8:39 AM
To: gopro/gopro-lib-node.gl <gopro-lib-node.gl@noreply.github.com>
Cc: Jeffrey Moguillansky <jmoguillansky@gopro.com>; Author <author@noreply.github.com>
Subject: Re: [gopro/gopro-lib-node.gl] Port to Windows / Visual Studio Toolchain on Master Branch (#152)


@cboesch-gpsw commented on this pull request.

________________________________

In libnodegl/utils.c<https://github.com/gopro/gopro-lib-node.gl/pull/152#discussion_r553443678>:

> +#define POW10_9 1000000000

+#include <Windows.h>

Yes but why re-ordering fix it? What includes set the architecture, and what is complaining about this missing architecture?

—
You are receiving this because you authored the thread.
Reply to this email directly, view it on GitHub<https://github.com/gopro/gopro-lib-node.gl/pull/152#discussion_r553443678>, or unsubscribe<https://github.com/notifications/unsubscribe-auth/AO3BUE55QDPGOOWOOZWX5HTSYXPR5ANCNFSM4UDAE7OA>.

jmoguillansky-gpsw commented:
gitattributes(5) (kernel.org)<https://mirrors.edge.kernel.org/pub/software/scm/git/docs/gitattributes.html>

text
This attribute enables and controls end-of-line normalization. When a text file is normalized, its line endings are converted to LF in the repository. To control what line ending style is used in the working directory, use the eol attribute for a single file and the core.eol configuration variable for all text files. Note that setting core.autocrlf to true or input overrides core.eol (see the definitions of those options in git-config(1)<https://mirrors.edge.kernel.org/pub/software/scm/git/docs/git-config.html>).

We’re disabling this feature (the minus in front) so we’re disabling end-of-line normalization. It won’t affect any binaries, etc.  All this does is turn off end-of-line normalization of text files.


From: cboesch-gpsw <notifications@github.com>
Sent: Thursday, January 7, 2021 8:40 AM
To: gopro/gopro-lib-node.gl <gopro-lib-node.gl@noreply.github.com>
Cc: Jeffrey Moguillansky <jmoguillansky@gopro.com>; Author <author@noreply.github.com>
Subject: Re: [gopro/gopro-lib-node.gl] Port to Windows / Visual Studio Toolchain on Master Branch (#152)


@cboesch-gpsw commented on this pull request.

________________________________

In .gitattributes<https://github.com/gopro/gopro-lib-node.gl/pull/152#discussion_r553444555>:

> @@ -0,0 +1 @@

+* -text

Also, doesn't this affect all binaries? Like, is git going to consider our images in the doc as text now?

—
You are receiving this because you authored the thread.
Reply to this email directly, view it on GitHub<https://github.com/gopro/gopro-lib-node.gl/pull/152#discussion_r553444555>, or unsubscribe<https://github.com/notifications/unsubscribe-auth/AO3BUEYEXTIMJKCJHRZECALSYXPXLANCNFSM4UDAE7OA>.

cboesch-gpsw commented on Makefile:
```diff
@@ -89,14 +124,26 @@ all: ngl-tools-install pynodegl-utils-install
 	@echo
 	@echo "    You can now enter the venv with:"
+ifeq ($(TARGET_OS),Windows)
+	@echo "        . $(PREFIX)\\Scripts\\activate.bat"
```
this file doesn't exist

jmoguillansky-gpsw replied:
It exists on windows (with msvc)

ubitux replied:
My bad I misunderstood.

cboesch-gpsw commented on Makefile:
```diff
@@ -19,36 +19,71 @@
+SET_PKG_CONFIG_PATH = PKG_CONFIG_PATH="$(W_PREFIX)\Lib\pkgconfig" WSLENV=PKG_CONFIG_PATH/w
+CMD = $(SET_PKG_CONFIG_PATH) cmd.exe /C
+else
+ACTIVATE = $(PREFIX)/bin/activate
```
why not integrate `cmd.exe` into `ACTIVATE` to help the different if branches?

jmoguillansky-gpsw replied:
cmd is running more commands than just activate.  It's doing activate and meson setup and meson compile, etc.  Putting cmd inside activate is a bit misleading

ubitux replied:
`CMD = .` then?

jmoguillansky-gpsw replied:
sure I'll update

jmoguillansky-gpsw replied:
let me know if this is better: https://github.com/gopro/gopro-lib-node.gl/pull/152/commits/7ca08730e1720b34116ee94f0e5418e9a91fc3fc

cboesch-gpsw commented on Makefile:
```diff
@@ -89,14 +124,26 @@ all: ngl-tools-install pynodegl-utils-install
 
 ngl-tools-install: nodegl-install
+ifeq ($(TARGET_OS),Windows)
+	($(CMD) $(ACTIVATE) \&\& $(MESON_SETUP) ngl-tools builddir\\ngl-tools \&\& $(MESON_COMPILE) -C builddir\\ngl-tools \&\& $(MESON_INSTALL) -C builddir\\ngl-tools)
```
why do you need to escape the `&`?

jmoguillansky-gpsw replied:
good question.  This is one of the tricky integration issues.  I'm passing ampersand to be processed by cmd executable instead of by bash shell (hence using the escape character).  I had to deal with various issues regarding slashes and ampersands etc.

jmoguillansky-gpsw commented:
It exists on windows

Get Outlook for Android<https://aka.ms/ghei36>

________________________________
From: cboesch-gpsw <notifications@github.com>
Sent: Thursday, January 7, 2021 8:53:05 AM
To: gopro/gopro-lib-node.gl <gopro-lib-node.gl@noreply.github.com>
Cc: Jeffrey Moguillansky <jmoguillansky@gopro.com>; Author <author@noreply.github.com>
Subject: Re: [gopro/gopro-lib-node.gl] Port to Windows / Visual Studio Toolchain on Master Branch (#152)


@cboesch-gpsw commented on this pull request.

________________________________

In Makefile<https://github.com/gopro/gopro-lib-node.gl/pull/152#discussion_r553452156>:

>       @echo

 ngl-tools-install: nodegl-install
+ifeq ($(TARGET_OS),Windows)
+       ($(CMD) $(ACTIVATE) \&\& $(MESON_SETUP) ngl-tools builddir\\ngl-tools \&\& $(MESON_COMPILE) -C builddir\\ngl-tools \&\& $(MESON_INSTALL) -C builddir\\ngl-tools)


why do you need to escape the &?

—
You are receiving this because you authored the thread.
Reply to this email directly, view it on GitHub<https://github.com/gopro/gopro-lib-node.gl/pull/152#pullrequestreview-563651325>, or unsubscribe<https://github.com/notifications/unsubscribe-auth/AO3BUEY5L7FYSYALWZADRJTSYXRHDANCNFSM4UDAE7OA>.

jmoguillansky-gpsw commented:
No what it means is that if you submit a text file with /n on windows, git will convert to /r/n.  We don't want git changing the line endings we can manage them ourselves

Get Outlook for Android<https://aka.ms/ghei36>

________________________________
From: Clément Bœsch <notifications@github.com>
Sent: Friday, January 8, 2021 12:37:38 AM
To: gopro/gopro-lib-node.gl <gopro-lib-node.gl@noreply.github.com>
Cc: Jeffrey Moguillansky <jmoguillansky@gopro.com>; Author <author@noreply.github.com>
Subject: Re: [gopro/gopro-lib-node.gl] Port to Windows / Visual Studio Toolchain on Master Branch (#152)


@ubitux commented on this pull request.

________________________________

In Makefile<https://github.com/gopro/gopro-lib-node.gl/pull/152#discussion_r553808615>:

> @@ -225,13 +225,25 @@ else
 endif

 tests: nodegl-tests tests-setup
+ifeq ($(TARGET_OS),Windows)
+       (cmd.exe /C $(ACTIVATE) \&\& meson test $(MESON_TESTS_SUITE_OPTS) -C builddir\\tests)


we use different slashes (e.g. forward slash instead of backslashes).

I though windows was supporting / nowadays?

It's not so easy to do this in a makefile, easier in a high level programming language like Python, etc.

I'm not opposed to having Windows handled outside the Makefile (we moved away from make for the build for a reason). I don't know if Python would be a great language for bootstrapping packages etc on Windows (because it requires an initial installation for the user and we will need to create a virtual env from another Python script, it might get clumsy), but I wouldn't be particularly opposed to that. Isn't Powershell moer appropriate? (I really don't know).

If you were under the impression that I was asking for Windows integration within the Makefile, I'm sorry. The goal of that Makefile is to bootstrap the environment: we kept make because on Unix it's extremely convenient. The keyword here is "convenient", so the question to answer on windows is: "what would be the most convenient way for a user to bootstrap the environment reliably?".

—
You are receiving this because you authored the thread.
Reply to this email directly, view it on GitHub<https://github.com/gopro/gopro-lib-node.gl/pull/152#discussion_r553808615>, or unsubscribe<https://github.com/notifications/unsubscribe-auth/AO3BUEZPPJEKY5LD3I2GL3LSY275FANCNFSM4UDAE7OA>.

jmoguillansky-gpsw commented:
Regarding supporting backslashes: not command prompt and I'm not familiar with powershell.  I used bash and Makefile because its the minimal porting effort.

Get Outlook for Android<https://aka.ms/ghei36>
________________________________
From: Clément Bœsch <notifications@github.com>
Sent: Friday, January 8, 2021 12:37:38 AM
To: gopro/gopro-lib-node.gl <gopro-lib-node.gl@noreply.github.com>
Cc: Jeffrey Moguillansky <jmoguillansky@gopro.com>; Author <author@noreply.github.com>
Subject: Re: [gopro/gopro-lib-node.gl] Port to Windows / Visual Studio Toolchain on Master Branch (#152)


@ubitux commented on this pull request.

________________________________

In Makefile<https://github.com/gopro/gopro-lib-node.gl/pull/152#discussion_r553808615>:

> @@ -225,13 +225,25 @@ else
 endif

 tests: nodegl-tests tests-setup
+ifeq ($(TARGET_OS),Windows)
+       (cmd.exe /C $(ACTIVATE) \&\& meson test $(MESON_TESTS_SUITE_OPTS) -C builddir\\tests)


we use different slashes (e.g. forward slash instead of backslashes).

I though windows was supporting / nowadays?

It's not so easy to do this in a makefile, easier in a high level programming language like Python, etc.

I'm not opposed to having Windows handled outside the Makefile (we moved away from make for the build for a reason). I don't know if Python would be a great language for bootstrapping packages etc on Windows (because it requires an initial installation for the user and we will need to create a virtual env from another Python script, it might get clumsy), but I wouldn't be particularly opposed to that. Isn't Powershell moer appropriate? (I really don't know).

If you were under the impression that I was asking for Windows integration within the Makefile, I'm sorry. The goal of that Makefile is to bootstrap the environment: we kept make because on Unix it's extremely convenient. The keyword here is "convenient", so the question to answer on windows is: "what would be the most convenient way for a user to bootstrap the environment reliably?".

—
You are receiving this because you authored the thread.
Reply to this email directly, view it on GitHub<https://github.com/gopro/gopro-lib-node.gl/pull/152#discussion_r553808615>, or unsubscribe<https://github.com/notifications/unsubscribe-auth/AO3BUEZPPJEKY5LD3I2GL3LSY275FANCNFSM4UDAE7OA>.

jmoguillansky-gpsw commented:
I don't think we should have makefile for one platform and something else for other platforms, ideally we use the same thing for both.  I'm ok with Makefile.

Get Outlook for Android<https://aka.ms/ghei36>
________________________________
From: Jeffrey Moguillansky <jmoguillansky@gopro.com>
Sent: Friday, January 8, 2021 7:27:18 AM
To: gopro/gopro-lib-node.gl <reply@reply.github.com>; gopro/gopro-lib-node.gl <gopro-lib-node.gl@noreply.github.com>
Cc: Author <author@noreply.github.com>
Subject: Re: [gopro/gopro-lib-node.gl] Port to Windows / Visual Studio Toolchain on Master Branch (#152)

Regarding supporting backslashes: not command prompt and I'm not familiar with powershell.  I used bash and Makefile because its the minimal porting effort.

Get Outlook for Android<https://aka.ms/ghei36>
________________________________
From: Clément Bœsch <notifications@github.com>
Sent: Friday, January 8, 2021 12:37:38 AM
To: gopro/gopro-lib-node.gl <gopro-lib-node.gl@noreply.github.com>
Cc: Jeffrey Moguillansky <jmoguillansky@gopro.com>; Author <author@noreply.github.com>
Subject: Re: [gopro/gopro-lib-node.gl] Port to Windows / Visual Studio Toolchain on Master Branch (#152)


@ubitux commented on this pull request.

________________________________

In Makefile<https://github.com/gopro/gopro-lib-node.gl/pull/152#discussion_r553808615>:

> @@ -225,13 +225,25 @@ else
 endif

 tests: nodegl-tests tests-setup
+ifeq ($(TARGET_OS),Windows)
+       (cmd.exe /C $(ACTIVATE) \&\& meson test $(MESON_TESTS_SUITE_OPTS) -C builddir\\tests)


we use different slashes (e.g. forward slash instead of backslashes).

I though windows was supporting / nowadays?

It's not so easy to do this in a makefile, easier in a high level programming language like Python, etc.

I'm not opposed to having Windows handled outside the Makefile (we moved away from make for the build for a reason). I don't know if Python would be a great language for bootstrapping packages etc on Windows (because it requires an initial installation for the user and we will need to create a virtual env from another Python script, it might get clumsy), but I wouldn't be particularly opposed to that. Isn't Powershell moer appropriate? (I really don't know).

If you were under the impression that I was asking for Windows integration within the Makefile, I'm sorry. The goal of that Makefile is to bootstrap the environment: we kept make because on Unix it's extremely convenient. The keyword here is "convenient", so the question to answer on windows is: "what would be the most convenient way for a user to bootstrap the environment reliably?".

—
You are receiving this because you authored the thread.
Reply to this email directly, view it on GitHub<https://github.com/gopro/gopro-lib-node.gl/pull/152#discussion_r553808615>, or unsubscribe<https://github.com/notifications/unsubscribe-auth/AO3BUEZPPJEKY5LD3I2GL3LSY275FANCNFSM4UDAE7OA>.

jmoguillansky-gpsw commented:
some recent updates:
1) added params --includedir and --libdir when calling meson setup to point to $(PREFIX)\Include and $(PREFIX)\Lib.  
This also fixes the pkg-config generation issue: now the generated pkg-config files point to correct location.
2) updated pkg-config patch:
set params "--keep-system-cflags and --keep-system-libs" automatically.  This resolves an issue where various pkg-config cflags and libs params were being filtered out by pkg-config (pkg-config thinks these directories are system paths but they're not).

cboesch-gpsw commented:
Merged with https://github.com/gopro/gopro-lib-node.gl/pull/177
